### PR TITLE
Per-port Destination rules and fault injection

### PR DESCRIPTION
**Pull request by rshriram**
Depends on https://github.com/envoyproxy/envoy/pull/3128 and
https://github.com/envoyproxy/envoy/pull/3084

Until then fault injection tests will fail.

Signed-off-by: Shriram Rajagopalan <shriramr@vmware.com>

### CODE REVIEW

File: pilot/pkg/networking/core/v1alpha3/route/route.go [hunk 1]
```diff
@@ -472,3 +479,49 @@ func BuildDefaultHTTPRoute(clusterName string) *route.Route {
...
+			Percent: uint32(in.Delay.Percent),
+		}
+		switch d := in.Delay.HttpDelayType.(type) {
+		case *networking.HTTPFaultInjection_Delay_FixedDelay:
+			delayDuration := util.GogoDurationToDuration(d.FixedDelay)
+			out.Delay.FaultDelaySecifier = &xdsfault.FaultDelay_FixedDelay{
+				FixedDelay: &delayDuration,
+			}
+		default:
+			log.Warnf("Ignoring exponential fault in route")
```
Comment (ZackButcher) [inline pilot/pkg/networking/core/v1alpha3/route/route.go]: `Exponential faults are not yet supported in routes`

or similar; when we say "ignoring" it's not clear if its the user's fault (e.g. bad config) or that it's not supported.

File: pilot/pkg/networking/core/v1alpha3/route/route.go [hunk 2]
```diff
@@ -472,3 +479,49 @@ func BuildDefaultHTTPRoute(clusterName string) *route.Route {
...
+		out.Abort = &xdshttpfault.FaultAbort{
+			Percent: uint32(in.Abort.Percent),
+		}
+		switch a := in.Abort.ErrorType.(type) {
+		case *networking.HTTPFaultInjection_Abort_HttpStatus:
+			out.Abort.ErrorType = &xdshttpfault.FaultAbort_HttpStatus{
+				HttpStatus: uint32(a.HttpStatus),
+			}
+		default:
+			log.Warnf("Ignoring non http status type faults")
```
Comment (ZackButcher) [inline pilot/pkg/networking/core/v1alpha3/route/route.go]: same as above:

`Currently only "status" type HTTP faults are supported in routes`

or similar.

File: pilot/pkg/networking/core/v1alpha3/route/route.go [hunk 3]
```diff
@@ -341,6 +344,11 @@ func translateRoute(in *networking.HTTPRoute,
 			}
 		}
 	}
+
+	if fault := in.Fault; fault != nil {
+		out.PerFilterConfig[xdsutil.Fault] = util.BuildProtoStruct(translateFault(in.Fault))
```
Comment (vadimeisenbergibm) [inline pilot/pkg/networking/core/v1alpha3/route/route.go]: `util.BuildProtoStruct` creates the struct inside "value" field, which produces the following configuration:
```
 "per_filter_config": {
          "envoy.fault": {
                   "value": {
                            "delay": {
                                "FaultDelaySecifier": {
                                       "FixedDelay": 7000000000
                                 },
                                 "percent": 100
                            }
                    }
        }
},
```

The "value" struct is redundant, its contents should appear immediately in the "envoy.fault" struct.

This causes the following error in Envoy:
`Error adding/updating listener http_0.0.0.0_9080: Unable to parse JSON as proto (INVALID_ARGUMENT:: Cannot find field.): {\"value\":{\"delay\":{\"FaultDelaySecifier\":{\"FixedDelay\":7000000000},\"percent\":100}}}"`

Comment (rshriram) [reply pilot/pkg/networking/core/v1alpha3/route/route.go]: Excellent find!!

**Comment by vadimeisenbergibm** [comment]
The current error:
```
Error adding/updating listener http_0.0.0.0_9080: Unable to parse JSON as proto (INVALID_ARGUMENT:delay: Cannot find field.): {\"delay\":{\"FaultDelaySecifier\":{\"FixedDelay\":7000000000},\"percent\":100}}
```

The configuration:
```json
"per_filter_config": {
                                                  "envoy.fault": {
                                                        "delay": {
                                                              "FaultDelaySecifier": {
                                                                    "FixedDelay": 7000000000
                                                                  },
                                                              "percent": 100
                                                            }
                                                      }
                                                },
```

**Comment by vadimeisenbergibm** [comment]
According to this code in Envoy https://github.com/envoyproxy/envoy/blob/d55f4990889da5d6874ab70ec5f4e5e0a9053160/source/common/config/utility.h#L239, per_filter_config has to contain JSONs.

```cpp
/*
* Translate a nested config into a route-specific proto message provided by
* the implementation factory.
* ...
*/
...
MessageUtil::jsonConvert(source, *config);
```

**Comment by vadimeisenbergibm** [comment]
The JSON schema for the HTTP Fault filter:
https://github.com/envoyproxy/envoy/blob/3b5fb0e6c8cbbaba1944cb8d74af3f35c471bde2/source/common/json/config_schemas.cc#L995

File: pilot/pkg/networking/core/v1alpha3/route/route.go (hunk 3 above)
Comment (vadimeisenbergibm) [reply pilot/pkg/networking/core/v1alpha3/route/route.go]: Found the problem, that line should be:
```
out.PerFilterConfig[xdsutil.Fault] = util.MessageToStruct(translateFault(in.Fault))
```

**Comment by rshriram** [comment]
merging as CLA bot is stuck.. All tests pass.

**Pull request merged**